<img src="http://oci02.img.iteso.mx/identidad_de_instancia_2018/ITESO/Logos%20ITESO/Logo-ITESO-Principal.jpg">

<center> <font color= #004A94> <font size = 6> Msc Ciencia de datos </font> <br> <br> <font color= #047CFB> <font size = 5>I.F. Juan Francisco Muñoz Elguezabal</font> <br> <br>

<center><font color= #42c29b> <font size = 6> Ejercicios varios de modelos lineales </font> <br> <br> <font color= #047CFB> <font size = 5>Repositorio: <a href='https://github.com/IFFranciscoME/FinTechLab/tree/master/'>Link</a></font>

---

## Cargar librerías y scritps

In [1]:
import numpy as np                                        # funciones matematicas
import pandas as pd                                       # dataframes y utilidades
import visualizaciones as vs                              # funciones de visualizacion
import statsmodels.api as sm                              # modelos estadisticos
from statsmodels.stats.multicomp import MultiComparison   # herramientas estadisticas: multicomparacion
from statsmodels.formula.api import ols                   # herramientas estadisticas: modelo lineal con ols
from sklearn.model_selection import train_test_split      # separar conjunto de entrenamiento y prueba
pd.options.mode.chained_assignment = None                 # para evitar el warning enfadoso de indexacion

## Ejercicio 1: Gasto Familiar

Se quiere ajustar un modelo que permita estimar los gastos en alimentación de una familia en base a la información que proporcionan las variables de ingresos mensuales y el número de miembros de la familia. Para ello se recoge una muestra aleatoria simple de 15 familiares cuyos resultados son los de la tabla adjunta. El gasto e ingresos está dado en cientos de miles de pesos

### Cargar y limpiar datos

In [2]:
df_gasto = pd.read_excel('archivos/datos_gastofamiliar.xlsx')

In [3]:
df_gasto.head()

Gasto  Ingreso  Tamaño
0   0.43      2.1       3
1   0.31      1.1       4
2   0.32      0.9       5
3   0.46      1.6       4
4   1.25      6.2       4

#### A) Obtenga el modelo lineal que permitirá estimar los gastos de alimentación en función de los ingresos mensuales de la familia.

In [4]:
x_train, x_test, y_train, y_test = train_test_split(df_gasto.loc[:, 'Ingreso'],
                                                    df_gasto.loc[:, 'Gasto'],
                                                    test_size=0.20, random_state=0, shuffle=False)

In [5]:
# Agregar interceptos a X en entrenamiento y prueba
x_train_betha = sm.add_constant(x_train)
x_test_betha = sm.add_constant(x_test)
modelo = sm.OLS(y_train, x_train_betha, hasconst=False)
resultados_a = modelo.fit()
print(resultados_a.summary())

                                 OLS Regression Results                                
Dep. Variable:                  Gasto   R-squared (uncentered):                   0.980
Model:                            OLS   Adj. R-squared (uncentered):              0.976
Method:                 Least Squares   F-statistic:                              241.7
Date:                Wed, 23 Oct 2019   Prob (F-statistic):                    3.42e-09
Time:                        13:13:59   Log-Likelihood:                          11.357
No. Observations:                  12   AIC:                                     -18.71
Df Residuals:                      10   BIC:                                     -17.74
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

/usr/local/lib/python3.6/dist-packages/scipy/stats/stats.py:1394: UserWarning:

kurtosistest only valid for n>=20 ... continuing anyway, n=12



#### B) ¿Cuáles el valor de las predicciones obtenidas por el modelo?

In [6]:
resultados_p = resultados_a.predict(np.array(x_test_betha))
df_resultados_p = pd.DataFrame({'Gasto_real': y_test,
                                'Gasto_pred': np.round(resultados_p, 2),
                                'Diff': np.round(resultados_p-y_test, 2)})
df_resultados_p.head()

Gasto_real  Gasto_pred  Diff
12        0.43        0.65  0.22
13        0.47        0.57  0.10
14        0.38        0.37 -0.01

#### C) Obtenga el modelo lineal que permitirá estimar los gastos de alimentación en función de los ingresos mensuales y del número de miembros de la familia.

In [7]:
x_train, x_test, y_train, y_test = train_test_split(df_gasto.loc[:, ['Ingreso','Tamaño']],
                                                    df_gasto.loc[:, 'Gasto'],
                                                    test_size=0.20, random_state=0, shuffle=False)
# Agregar interceptos a X en entrenamiento y prueba
x_train_betha = sm.add_constant(x_train)
x_test_betha = sm.add_constant(x_test)
modelo = sm.OLS(y_train, x_train_betha, hasconst=False)
resultados_a = modelo.fit()
print(resultados_a.summary())

                                 OLS Regression Results                                
Dep. Variable:                  Gasto   R-squared (uncentered):                   0.988
Model:                            OLS   Adj. R-squared (uncentered):              0.985
Method:                 Least Squares   F-statistic:                              257.3
Date:                Wed, 23 Oct 2019   Prob (F-statistic):                    4.88e-09
Time:                        13:13:59   Log-Likelihood:                          14.742
No. Observations:                  12   AIC:                                     -23.48
Df Residuals:                       9   BIC:                                     -22.03
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

#### D) ¿Cuáles son las predicciones obtenidas por el modelo?

In [8]:
resultados_p = resultados_a.predict(np.array(x_test_betha))
df_resultados_p = pd.DataFrame({'Gasto_real': y_test,
                                'Gasto_pred': np.round(resultados_p, 2),
                                'Diff': np.round(resultados_p-y_test, 2)})
df_resultados_p.head()

Gasto_real  Gasto_pred  Diff
12        0.43        0.55  0.12
13        0.47        0.52  0.05
14        0.38        0.36 -0.02

#### E) Obtenga los contrastes individuales de los coeficientes $\beta_i$ con $\alpha=0.05$  ¿Qué puede concluir con los resultados obtenidos?

| Parámetro | Coeficiente $\beta_i$  | p_value Vs $\alpha$      | Conclusión 1        |
| --------- | ---------------------- | ------------------------ | ------------------- |
| const     | -0.0944                | 0.412 > 0.05             | No es significativo |
| Ingreso   | 0.1476                 | 0.000 < 0.05             | Si es significativo | 
| Tamaño    | 0.0625                 | 0.028 < 0.05             | Si es significativo |

El ingreso de una familia puede ser explicado según su ingreso y su tamaño. En ambos casos, a mayor ingreso, mayor gasto y a mayor tamaño mayor gasto también.

#### F) Estime el gasto en alimentación de una familia con ingresos de 3.0 y un tamaño de 4.

In [9]:
x_test_betha_1 = pd.DataFrame({'const': 1, 'Ingreso': 3.0, 'Tamaño': 4}, index=[0])
resultados_p = resultados_a.predict(np.array(x_test_betha_1))

print('El gasto estimado de una familia con ingresos de: ' + str(x_test_betha_1['Ingreso'][0]) + 
      ' y de un tamaño de: ' + str(x_test_betha_1['Tamaño'][0]) + ' se pronostica será de: ' + 
      str(np.round(resultados_p[0])))

El gasto estimado de una familia con ingresos de: 3.0 y de un tamaño de: 4 se pronostica será de: 1.0


#### G) ¿Qué pasaría si se quisiera estimar el gasto en la alimentación de una familia con 8 integrantes?

Dejaré esto por aquí y continuaré con los otros ejercicios

<img src="https://imgs.xkcd.com/comics/extrapolating.png">

#### H) Buscar una base de datos para regresión lineal de la siguiente liga  https://archive.ics.uci.edu/ml/index.php  y  obtener diferentes modelos lineales y hacer sus respectivos contrastes de hipótesis.

## Ejercicio 2: Peces enfermos

Se quiere estudiar el efecto de distintas dosis de un medicamento para combatir a los parásitos de peces criados en acuicultura. Para ello, se tomaron 60 peces al azar, y se dividieron en 5 grupos de 12 individuos cada uno. El primer grupo no fue medicado, pero a los restantes se les suministró el medicamento en dosis crecientes. Tras una semana de tratamiento, se contabilizaron los parásitos existentes en cada individuo, obteniendo los resultados siguientes:

In [10]:
datos = pd.read_excel('archivos/datos_peces.xlsx', header=None)
datos.columns = ['grupo'] + ['pez_' + str(i) for i in range(1, len(datos.columns))]

In [11]:
datos

grupo  pez_1  pez_2  pez_3  pez_4  pez_5  pez_6  pez_7  pez_8  pez_9  \
0  Control     50     65     72     46     38     29     70     85     72   
1     25mg     49     47     30    602     62     60     19     28     56   
2     50mg     20     59     64     61     28     47     29     41     60   
3    100mg     20     23     38     31     27     16     27     18     22   
4    125mg     18     30     22     26     31     11     15     12     31   

   pez_10  pez_11  pez_12  
0      40      57      59  
1      62      55      40  
2      57      61      38  
3      12      24      11  
4      36      16      13

In [12]:
datos = datos.melt(id_vars='grupo')
datos = datos[['grupo', 'value']]
datos = datos.sort_values(by=['grupo'])
datos.head(14)

grupo  value
18  100mg     31
53  100mg     24
3   100mg     20
28  100mg     16
48  100mg     12
58  100mg     11
43  100mg     22
8   100mg     23
23  100mg     27
38  100mg     18
13  100mg     38
33  100mg     27
24  125mg     31
29  125mg     11

#### A) Contrastar si el medicamento es efectivo contra los parásitos con $\alpha=0.05$ y dar una interpretación de los resultados obtenidos.

In [13]:
# ajustar modelo lineal para (high - open)
modelo_peces = ols('value ~ C(grupo)', data=datos[['grupo', 'value']]).fit()
modelo_peces.summary()

# tabla anova
anova_table = sm.stats.anova_lm(modelo_peces, typ=2)
anova_table

sum_sq    df        F    PR(>F)
C(grupo)   40848.933333   4.0  1.91933  0.120162
Residual  292640.000000  55.0      NaN       NaN

### Planteamiento de Hipótesis

Estadísticamente:

- **H0**: la media de los grupos es la misma.
- **Ha**: la media de al menos un grupo es diferente de la de los demás grupos.

En el contexto del problema: 

- **H0**: Entre todos los grupos, no existe reducción en el promedio de parásitos encontrados en los peces.
- **Ha**: Al menos en 1 grupo hubo variación significativa en el promedio de parásitos encontrado en los peces. 

Se observa que el valor p es **0.120162** y es > 0.05, por lo tanto *se acepta* la **H0:  Entre todos los grupos, no hubo reducción en el promedio de parásitos encontrados en los peces después de aplicar diferentes tratamientos en cada grupo**.

No podríamos atribuir si es debido a la inefectividad del medicamento que no hubo reducción en los parásitos de los peces, tendría que hacerse un diagrama causal, pero, algo si podemos hacer. Podemos cantar, podemos cantar con tristeza pero con certeza estadística, *Pero mira como mueren los peces en el rio ... mueren y mueren los peces en el rio*

## Ejercicio 3: Guadalajara ya no huele a tierra mojada

In [1]:
datos = pd.read_excel('archivos/datos_ambientales.xlsx', header=None)

NameError: name 'pd' is not defined

## Ejercicio 4: Aspirina para el corazon

In [87]:
datos = pd.DataFrame({'Mortal': [18, 5], 'No mortal': [171, 99], 'No ataque': [10845, 10933]}, index=['Placebo', 'Aspirina'])
total_p = 11034
total_a = 11037

In [88]:
datos

Mortal  No mortal  No ataque
Placebo       18        171      10845
Aspirina       5         99      10933

#### Calcular los log de: $x_{ij}$

In [89]:
datos_log = np.log(datos)

In [90]:
datos_log

Mortal  No mortal  No ataque
Placebo   2.890372   5.141664   9.291459
Aspirina  1.609438   4.595120   9.299541

#### Calcular y colocar la media de los datos por columna $c$ y la media de los datos por fila $f$

In [91]:
datos_log.loc['media_c'] = datos_log.mean(axis=0)
datos_log['media_f'] = datos_log.mean(axis=1)
datos_log

Mortal  No mortal  No ataque   media_f
Placebo   2.890372   5.141664   9.291459  5.774498
Aspirina  1.609438   4.595120   9.299541  5.168033
media_c   2.249905   4.868392   9.295500  5.471266

In [98]:
# media global
mu = datos_log.iloc[-1,-1]
# media de las filas
mu1 = datos_log.iloc[:,-1]-mu
# media de las columnas
mu2 = datos_log.iloc[-1,:]-mu
mu2

Mortal      -3.221361
No mortal   -0.602874
No ataque    3.824235
media_f      0.000000
Name: media_c, dtype: float64

#### Reconstrucción la tabla de los logaritmos

In [93]:
datos_rec = np.ones(np.shape(datos_log))*mu
datos_rec

array([[5.47126559, 5.47126559, 5.47126559, 5.47126559],
       [5.47126559, 5.47126559, 5.47126559, 5.47126559],
       [5.47126559, 5.47126559, 5.47126559, 5.47126559]])

In [94]:
# sumarle a cada valor la media de las filas
for k in np.arange(np.shape(datos_log)[0]):
    datos_rec[k,:] = datos_rec[k,:]+mu2

# sumarle a cada valor la media de las columnas
for k in np.arange(np.shape(datos_log)[1]):
    datos_rec[:,k] = datos_rec[:,k]+mu1

# seleccionar solo los valores originales (sin filas de mu1 ni mu2)
datos_rec = datos_rec[0:-1, 0:-1]

In [95]:
datos_rec

array([[2.55313749, 5.17162436, 9.59873288],
       [1.94667218, 4.56515904, 8.99226756]])

In [96]:
tab_mu12 = pd.DataFrame(np.transpose(np.transpose(datos_log-mu-mu2)-mu1))
tab_mu12.iloc[:,-1] = mu1
tab_mu12.iloc[-1,:] = mu2
tab_mu12

Mortal  No mortal  No ataque   media_f
Placebo   0.337234  -0.029961  -0.307273  0.303233
Aspirina -0.337234   0.029961   0.307273 -0.303233
media_c  -3.221361  -0.602874   3.824235  0.000000

In [64]:
datos_rec_mod = datos_rec + tab_mu12.iloc[0:-1,0:-1]

In [65]:
datos_rec_mod

Mortal  No mortal  No ataque
Placebo   2.890372   5.141664   9.291459
Aspirina  1.609438   4.595120   9.299541

In [48]:
datos_log

Mortal  No mortal  No ataque   media_f
Placebo   2.890372   5.141664   9.291459  5.774498
Aspirina  1.609438   4.595120   9.299541  5.168033
media_c   2.249905   4.868392   9.295500  5.471266

Las variables son independientes, puesto que se cumple que la reconstruccion de la matriz de los logaritmos es igual a la matriz original, esto haciéndolo a partir de la suma del valor esperado de cada casilla con las medias de cada fila y cada columna